frontend/src/scenes/insights/InsightPageHeader.tsx

```diff
@@ -330,7 +330,14 @@ export function InsightPageHeader({ insightLogicProps }: { insightLogicProps: In
                             <LemonButton
                                 tooltip={showQueryEditor ? 'Hide JSON editor' : 'Edit as JSON'}
                                 type={'secondary'}
-                                onClick={toggleQueryEditorPanel}
+                                onClick={() => {
+                                    toggleQueryEditorPanel()
+                                    if (insightMode !== ItemMode.Edit) {
+                                        setInsightMode(ItemMode.Edit, null)
+                                    } else if (insightMode === ItemMode.Edit && insight.saved) {
```

Comment:
We shouldn't rely on the `saved` field, it's not guaranteed to mean anything now (used to mean something when every insight results request added an `Insight` row to the database, and saving was built on top of that)

Reply:
Hmm.. I'm not sure if I understand what you're saying (but maybe it's just too late for me today). Can I repro an insight without `saved` field somehow locally? Imo the cancel button that's also based on `insight.saved` wouldn't show up then as well.

Reply:
I'm now basing the check for insight persistence on the `hasDashboardItemId` selector. I've also changed the "..." more menu to be based on wether the insight is persisted instead of it being currently edited.